### PR TITLE
TF-223: When a type does not conform, explain why

### PR DESCRIPTION
**Pull request by marcrasi**
The LLDB diagnostic logic hides diagnostics pointing to things that the user did not enter. For example, if you do `struct Foo : Sequence {}`, then you should get this:
```
error: repl.swift:1:8: error: type 'Foo' does not conform to protocol 'Sequence'
struct Foo : Sequence {}
       ^

Swift.Sequence:3:20: note: unable to infer associated type 'Iterator' for protocol 'Sequence'
    associatedtype Iterator : IteratorProtocol
                   ^

Swift.Sequence:2:40: note: candidate would match and infer 'Iterator' = 'Foo' if 'Foo' conformed to 'IteratorProtocol'
    @inlinable public __consuming func makeIterator() -> Self
```
but LLDB hides the two notes because they don't point at the user's expression.

This PR fixes that.

I'm not sure if there are any situations where this fix breaks some desired behavior. I'll create a PR into master soon, and I'll ask in that PR whether there are such situations.

Fixes https://bugs.swift.org/browse/TF-223

### CODE REVIEW

File: source/Symbol/SwiftASTContext.cpp [hunk 1]
```diff
@@ -2807,55 +2807,61 @@ class StoringDiagnosticConsumer : public swift::DiagnosticConsumer {
...
+          fixed_description.Printf("%s:%u:",
+                                   diagnostic.bufferName.str().c_str(),
+                                   diagnostic.line - first_line +
+                                       line_offset + 1);
+          start_pos = match_pos + match_len;
+          match_pos =
+              diagnostic.description.find(match.GetString(), start_pos);
+        } while (match_pos != std::string::npos);
+
+        // Append any last remainging text
```
Comment (bgogul) [inline source/Symbol/SwiftASTContext.cpp:2846]: typo: remainging -> remaining

(The old code has the typo too.)

**Review by bgogul — APPROVED** [review]

**Comment by jimingham** [comment]
The problem is, for instance, every time an expression fails to parse we get a bunch of warnings that have nothing to do with the compile failure (e.g. that we didn't use the variable we use to capture errors.  Those are just confusing, so we suppress them.

**Comment by marcrasi** [comment]
I think that this change handles that case, because I have made it still exclude errors from the expression's buffer that fall outside of `[first_line, last_line]`. So if those confusing warnings all appear in the same buffer as the expression, they should still be excluded.

**Comment by jimingham** [comment]
That should be fine, then.  Avoiding printing errors/warnings for the implementation details of the expression parser was the whole point of this exercise.

**Pull request merged**